src/main/java/org/truffleruby/extra/RubyConcurrentHashMap.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2015, 2020 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.extra;
+
+import com.oracle.truffle.api.object.Shape;
+import org.truffleruby.RubyContext;
+import org.truffleruby.core.basicobject.RubyBasicObject;
+import org.truffleruby.core.klass.RubyClass;
+import org.truffleruby.language.RubyDynamicObject;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+public class RubyConcurrentHashMap extends RubyDynamicObject {
+
+   public static class Key {
+
+       public final Object key;
+
+       public Key(Object key) {
+           this.key = key;
+       }
+
+       @Override
+       public int hashCode() {
+           final Object code = RubyContext.send(key, "hash");
+           if (code instanceof Integer) {
+               return (int) code;
+           } else if (code instanceof Long) {
+               return (int) (long) code;
+           } else {
+               throw new UnsupportedOperationException(code.getClass().getName());
+           }
+       }
+
+       @Override
+       public boolean equals(Object otherKey) {
+           assert otherKey instanceof Key;
+           final Object code = RubyContext.send(key, "eql?", ((Key) otherKey).key);
```

Comment:
Use `SameOrEqlNode.getUncached().execute()` here.
We should probably also do the optimization of `CompareHashKeysNode` which first compares the hashCode since that's a cheap check which might short-circuit early if the keys are different and avoid the actual `eql?` call.

Reply:
> We should probably also do the optimization of CompareHashKeysNode which first compares the hashCode since that's a cheap check which might short-circuit early if the keys are different and avoid the actual eql? call.

We have the short-circuit check for `hashCode` now: https://github.com/oracle/truffleruby/pull/2339/files#diff-bb88eeac5e669b5e6f55477c4c61094b8d85c463e546f15bb7df4ffd889a8d8eR41

> Use SameOrEqlNode.getUncached().execute() here.

I created a new function in KernelNodes. 
```
public static SameOrEqlNode getUncached() { 
            return KernelNodesFactory.SameOrEqlNodeGen.getUncached(); 
        }
```

And then used it as so, `final Object returnValue = KernelNodes.SameOrEqlNode.getUncached().execute(key, otherKey.key);` But this gives me a `SystemStackError` when I run the specs:

```
SystemStackError: stack level too deep
	from KernelNodes.java:233:in `org.truffleruby.core.kernel.KernelNodes$SameOrEqlNode.getUncached'
/Users/mapleong/src/github.com/Shopify/truffleruby/spec/truffle/truffleruby/concurrent_hash_map_spec.rb:136:in `get_or_default'
/Users/mapleong/src/github.com/Shopify/truffleruby/spec/truffle/truffleruby/concurrent_hash_map_spec.rb:136:in `block (2 levels) in <top (required)>'
/Users/mapleong/src/github.com/Shopify/truffleruby/spec/truffle/truffleruby/concurrent_hash_map_spec.rb:3:in `<top (required)>'
/Users/mapleong/src/github.com/Shopify/truffleruby/spec/ruby/spec_helper.rb:31:in `<top (required)>'
```

Am I missing something?

Reply:
SameOrEqlNode is `@GenerateUncached` so it should work, but you probably need to rebase to latest `master`

Reply:
FYI you can clock Ctrl+click on `getUncached()` there to see where it ends up, after `jt build` to make sure the generated sources are up to date.

Reply:
Rebasing fixed the issue - [`0ab563b` (#2339)](https://github.com/oracle/truffleruby/pull/2339/commits/0ab563bc99b7f40bf5027bf930e3daadfce3862d)